lib/Sema/TypeCheckPattern.cpp

```diff
@@ -727,9 +727,11 @@ validateTypedPattern(TypedPattern *TP, DeclContext *dc,
   // property definition.
   auto &Context = dc->getASTContext();
   auto *Repr = TP->getTypeRepr();
-  if (Repr && Repr->hasOpaque()) {
+  if (Repr && (Repr->hasOpaque() ||
+               (Context.LangOpts.hasFeature(Feature::ImplicitSome) &&
+                Repr->isProtocol(dc)))) {
```

Comment:
It's not immediately obvious to me why we need to check for the flag here when we check in `TypeCheckType.cpp` ?

Reply:
Good question - outside this condition is a call to `TypeResolution::resolveContextualType`, which will go into type resolution without having set the opaque type decl, so it was resolving protocols in result position as existential types since `dyn_cast<OpaqueTypeDecl>(DC)` returned `nullptr` in that case. I agree it's unfortunate to check the flag in so many places though, perhaps there's a way to factor most of the logic into one place...